src/main/java/com/linecorp/armeria/common/RequestContext.java

```diff
@@ -0,0 +1,290 @@
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.FutureListener;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+
+/**
+ * Provides information about an invocation and related utilities. Every remote invocation, regardless of if
+ * it's client side or server side, has its own {@link RequestContext} instance.
+ */
+public interface RequestContext extends AttributeMap {
+
+    /**
+     * Returns the context of the invocation that is being handled in the current thread.
+     *
+     * @throws IllegalStateException if the context is unavailable in the current thread
+     */
+    static <T extends RequestContext> T current() {
+        final T ctx = RequestContextThreadLocal.get();
+        if (ctx == null) {
+            throw new IllegalStateException(RequestContext.class.getSimpleName() + " unavailable");
+        }
+        return ctx;
+    }
+
+    /**
+     * Maps the context of the invocation that is being handled in the current thread.
+     *
+     * @param mapper the {@link Function} that maps the invocation
+     * @param defaultValueSupplier the {@link Supplier} that provides the value when the context is unavailable
+     *                             in the current thread. If {@code null}, the {@code null} will be returned
+     *                             when the context is unavailable in the current thread.
+     */
+    static <T> T mapCurrent(
+            Function<? super RequestContext, T> mapper, @Nullable Supplier<T> defaultValueSupplier) {
+
+        final RequestContext ctx = RequestContextThreadLocal.get();
+        if (ctx != null) {
+            return mapper.apply(ctx);
+        }
+
+        if (defaultValueSupplier != null) {
+            return defaultValueSupplier.get();
+        }
+
+        return null;
+    }
+
+    /**
+     * (Do not use; internal use only) Sets the invocation context of the current thread.
+     */
+    static PushHandle push(RequestContext ctx) {
+        requireNonNull(ctx, "ctx");
+        final RequestContext oldCtx = RequestContextThreadLocal.getAndSet(ctx);
+        return oldCtx != null ? () -> RequestContextThreadLocal.set(oldCtx)
+                              : RequestContextThreadLocal::remove;
+    }
+
+    /**
+     * Returns the {@link SessionProtocol} of this request.
+     */
+    SessionProtocol sessionProtocol();
+
+    String method();
+
+    /**
+     * Returns the absolute path part of this invocation, as defined in
+     * <a href="http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2">the
+     * section 5.1.2 of RFC2616</a>.
+     */
+    String path();
+
+    /**
+     * Returns the request associated with this context.
+     */
+    <T> T request();
+
+    RequestLogBuilder requestLogBuilder();
+    ResponseLogBuilder responseLogBuilder();
+    CompletableFuture<RequestLog> awaitRequestLog();
+    CompletableFuture<ResponseLog> awaitResponseLog();
```

Comment:
note: It seems like logging features are deeply bound to overall request context. i.e. they will be generated regardless of logging decorator(or any listener) exists or not.

but I have no idea how to improve this impl. with preserving current feature set.